### PR TITLE
perf(pathways): reduce per-step overhead; fix PD abort lifecycle, cpp fastpath and FP8 scale sharding

### LINKED ISSUE #1452: [perf] jitted_run_model cpp fastpath miss every decode step under dp=1 + Pathways proxy backend
# [perf] `jitted_run_model` cpp fastpath miss every decode step under dp=1 + Pathways proxy backend

## Symptom

On GLM-5.2-FP8 (epmoe/MLA) with `--dp-size=1` under Pathways single-controller (`JAX_PLATFORMS=proxy`), every warm decode step reports `#cache_miss: 1` from `count_pjit_cpp_cache_miss()` around `jitted_run_model`. Observed on both **colocated** and PD-disaggregated configs — this is not PD-specific.

Not observed with `--dp-size=2` (MiMo-V2 tp8 dp2: `#cache_miss: 0` in steady state).

Overhead: ~<1ms/step (<1.1% of 89ms ITL on GLM tp16), so not a throughput blocker, but it defeats the pjit cpp fastpath and forces the Python slow path on every call.

## Root cause

`ForwardBatch.from_model_worker_batch` (via `device_array`) constructs a **fresh `NamedSharding` object per call** for each metadata array (`input_ids`, `positions`, `seq_lens`, `cache_loc`, ...). Instrumented `id(sharding)` differs every step; `id(sharding.mesh)` is stable.

jaxlib `PjitFunctionCache` key (`CallSignature`, `jaxlib/jax_jit.{h,cc}` @ jax-v0.8.1):
- `HashShardingForJit(NamedSharding)` = `absl::Hash<void*>()(mesh().ptr())` — stable ✓
- `EqualShardingsForJit(NamedSharding a, b)`:
  - fast-path `a.ptr() == b.ptr()` — **fails** (fresh object each call)
  - slow-path compares `mesh().ptr()` (stable) + `spec` (stable) + `memory_kind.equal()` + `logical_device_ids().equal()`

One of `memory_kind` / `logical_device_ids` / `layout` (also in `CallSignature`) differs across fresh `NamedSharding` objects under dp=1 (size-1 `data` axis). Under dp=2 the same fields are stable, so the slow-path `==` returns true and the cache hits.

Verified: `_get_fastpath_data` returns non-None (so it's not an output-type or effects issue); `JAX_EXPLAIN_CACHE_MISSES` shows tracing-cache hits after the first per-bs call — the miss is purely the cpp `PjitFunctionCache` lookup.

## Repro

```bash
python -m sgl_jax.launch_server \
  --model-path <GLM-5.2-FP8> --tp-size 16 --ep-size 16 --dp-size 1 \
  --page-size 128 --attention-backend fa --moe-backend epmoe \
  --disable-radix-cache --mem-fraction-static 0.86 \
  --precompile-bs-paddings 1 4 8 16 32 --precompile-token-paddings 4096
# any load; every "Decode batch" log line shows "#cache_miss: 1"
```

## Proposed fix

Cache and reuse `NamedSharding` objects in `device_array` (or at the `ForwardBatch`/`ModelWorkerBatch` level) keyed on `(mesh, spec)`, so `sharding.ptr()` is stable across calls and `EqualShardingsForJit`'s fast-path `a.ptr() == b.ptr()` short-circuits true. This also avoids the per-call `NamedSharding.__init__` overhead.

Alternative: pin `in_shardings` on `jitted_run_model` so jax reshards inputs to a fixed sharding object before the cache lookup.

## Environment

- jax 0.8.1, Pathways proxy backend (`JAX_PLATFORMS=proxy`)
- TPU multi-host, 1 slice tp16 dp1
- sglang-jax main @ 0381ccbf

### PR DESCRIPTION
**Pull request by zorrofox**
Refs #772, #1486; Fixes #1452, #1499

### Motivation

This is the Stage-4 batch of the Pathways integration roadmap item ("Benchmark and optimize Pathways integration, including the remaining device_get/host-transfer overhead and the decode fast-path miss" — #772, #774, #1452), plus the async P-pipeline lifecycle gap reported in #1486.

We first reproduced the #772 gap with DeepSeek-R1-Distill-Qwen-1.5B (tp4, dp1, bf16, page 256, chunked prefill 2048, radix off; `bench_serving` random 1024/1024, 384 prompts, max-concurrency 256; 3 rounds, warm rounds quoted, variance <1.5%) on identical single-host TPU v7x hardware, native vs `--device=proxy --enable-single-process`, and decomposed each decode step into H2D metadata upload / dispatch / resolve-D2H. The entire gap sits in the blocking `jax.device_get(next_token_ids)` in `resolve_last_batch_result`: ~21 ms per step for a few-hundred-byte array under the proxy backend (pure client→proxy→worker round-trip), vs ~3 ms native. H2D and dispatch are at parity.

Numbers on the current 0.10.x dependency set (see the compatibility note on Pathways image tags):

| config | stack (all jax 0.10.2) | output tok/s | median ITL |
|---|---|---|---|
| native (main baseline) | client 0.10.2 | 9682 (9662–9695 ×3) | 29.9 ms |
| native (this PR) | client 0.10.2 | 9675 | 29.8 ms |
| Pathways (main baseline) | client + Pathways images 0.10.2 | 4047 | 49.5 ms |
| Pathways (this PR) | client + Pathways images 0.10.2 | 5344 | 38.8 ms |

(For reference, on the previously-validated jax 0.8.1 stack the same A/B was native 8987 vs Pathways 3938 → 5365 tok/s, i.e. 43.8% → 59.8% of native.)

### Compatibility notes: first full validation on the jax 0.10.x stack

Pathways serving had previously only been validated up to the jax 0.9 series; this PR's validation is the first on 0.10.x. Two pitfalls worth documenting:

1. **Pathways image tags.** The **static** image tags (`pathways/{server,proxy_server}:jax-0.10.1`, and there is no static `jax-0.10.2` tag) point at older binaries whose Mosaic deserializer tops out at version 13, so a jax 0.10.2 client fails every Pallas kernel under `--device=proxy` with `Failed to deserialize the Mosaic module: expected <= 13 but got 15` (native is unaffected — compile and execute share a process). The **dated daily builds** (e.g. `:20260727-jax_0.10.2`, also tagged `:latest`) are a single multi-version binary and work correctly with the 0.10.2 pin — all Pathways numbers and regression runs in this PR use `20260727-jax_0.10.2`, i.e. the exact pinned dependency set end to end.

2. **FP8 block-quant models crash on 0.10.x without the #1499 fix.** jax 0.8.x `shard_map` silently reshards on a textual in_specs mismatch; 0.10.x validates strictly. Static-quant checkpoints (GLM-5.2-FP8 and other DeepSeek-format FP8 models) load their compact 2D linear scales, expand them 2D→3D, and assign without resharding to the placeholder's declared sharding — so on 0.10.x every such model crashes at the first precompile, native and proxy alike (reproduced on unmodified main both ways; #1493 fixed the same class at the MoE/MLA boundaries but missed the linear path). Fixed here by resharding the expanded scale to the placeholder's sharding (see #1499 for the full analysis, including a static audit of the remaining latent spots — none of which trigger at tp16).

### Changes

1. **Start the next-token D2H on the forward thread** — right after the replicate-gather, `next_token_ids.copy_to_host_async()` so the scheduler's `resolve_last_batch_result` finds the copy in flight instead of paying the full interactive round-trip. Decode-phase resolve drops 21→8 ms/step under proxy (+33% output throughput); native unaffected (3 rounds, within variance).
2. **`device_array`: canonical `NamedSharding` per (mesh, spec, memory_kind)** — pjit's cpp fast-path compares input shardings by pointer; constructing a fresh `NamedSharding` per call defeats it and pays `NamedSharding.__init__` per metadata array per step (+2.5% under proxy on the same config).
3. **Async P-pipeline abort/retract lifecycle (#1486)** — the in-flight rid set becomes a registry `rid → (req, p_idx, reserved tokens, outstanding batch count, pending_abort)`. `abort_request` reaches requests anywhere in the pipeline: idle entries finalize immediately (P KV + req slot + admission token accounting + client abort, exactly once), in-flight entries finalize when their batch drains; chunk continuation stops for aborted owners; requests that already migrated are flagged finished so the standard prefill-result path frees their D KV without touching the mixed batch's page layout; pause/retract quiesces the pipeline so a late P result can't merge beside a requeued copy. The two `scheduler.py` hooks are getattr-guarded no-ops outside pathways PD.
4. **Replace `jax.ShapeDtypeStruct` state placeholders (#1452)** — static-quant checkpoints leave SDS placeholders in module state; jaxlib's `ToPyArgSignature` rejects them (`ComputeCallSignature failed: INVALID_ARGUMENT ... Got type ShapeDtypeStruct`, visible with `TF_CPP_VMODULE=pjit=3,jax_jit=3`), so every `jitted_run_model` call takes the python dispatch path — the per-decode-step cpp cache miss reported in #1452. The placeholders are dead on the forward path (a used leaf with a changed shape would fail to compile); they're swapped for zero-length arrays at `initialize_jit`, with the replaced paths logged. Note the root cause is unrelated to dp/proxy as originally guessed in the issue — the original attribution was a control-group artifact.
5. **Reshard expanded block-quant linear scales to their declared sharding (#1499)** — unblocks every FP8 block-quant checkpoint on jax 0.10.x (crash at first precompile on unmodified main, native and proxy alike); details in the compatibility note above and in #1499.

### Validation

- unit: sharding canonicalization + SDS-fastpath repro/restore (`test/srt/test_jax_utils.py`), lifecycle registry/abort/quiesce (`test/srt/disaggregation/test_pathways_pd.py`, 7 new tests)
- #772 repro: numbers above; greedy outputs byte-identical native vs proxy; gsm8k 200q native 0.600 / proxy 0.600
- #1486 e2e (MiMo tp8 2P1D Pathways): 24 aborts at t=2s → 24/24 terminated, registry marks confirmed in logs for in-pipeline requests, post-abort health 8/8, no leak after 12-abort stress round, 0 tracebacks
- #1452 (GLM static-quant MLA/epmoe tp16 dp1 + proxy): 156 placeholders replaced; decode `#cache_miss` 545/545 → 0 in steady state (only legit first-call-per-shape misses remain); gsm8k before/after 0.820 / 0.840 (parity)
- PD regression matrix (MiMo SWA/GQA tp8 2P1D, Pathways), run on BOTH stacks: jax 0.8.1 — correctness 336/336, bench 15534/15486/15512 tok/s (baseline band), gsm8k 0.855, no crash; jax 0.10.2 end-to-end (first 0.10.x Pathways-PD validation) — correctness 336/336, bench 15778/15713/15769 tok/s, gsm8k 0.890, no crash
- GLM-5.2-FP8 (MLA/epmoe) PD matrix @0.10.2 (tp16 2-slice, Pathways; requires the #1499 fix to even start on this stack): correctness 336/336 (SHORT/MID/LONG × conc 4/8/16), bench C64 16K/4K 192-req — 1276.0 / 1274.7 tok/s (r1/r2, 0.1% variance), median ITL 75.08 / 75.18 ms, no crash. The pre-PR baseline on the same workload/topology (jax 0.8.1 stack) was 1043.6 tok/s at 89.28 ms median ITL — the −14.2 ms/step matches the decode-resolve RTT this PR removes (21→8 ms measured in the #772 decomposition), plus the 0.10.2 stack gain. gsm8k 200q: 0.965 (a second independent bring-up scored 0.955).
- #1499 fix evidence: unmodified main crashes at first precompile on GLM static-quant, reproduced BOTH under `--device=proxy` (PD) and native `--device=tpu` (identical `shard_map in_specs ... does not match` error); with the fix the same configs fire up and pass (native: bench 1024/1024 C256 ×3 rounds = 1106.1 / 1106.1 / 1106.0 output tok/s, <0.01% variance, no crash).

### CODE REVIEW

File: python/sgl_jax/srt/disaggregation/pathways_scheduler.py [hunk 1]
```diff
@@ -654,6 +677,10 @@ def _pd_drain_ready(self) -> None:
             t_enq_ready,
         ) = item
         p_alloc, p_r2t, p_tree = self.p_allocs[p_idx], self.p_r2ts[p_idx], self.p_trees[p_idx]
+        # This batch is no longer in flight for its reqs (#1486); any
+        # pending_abort entry that reaches outstanding == 0 below is
+        # finalized on the matching path (done / mid-chunk / failure).
+        self._pd_batch_outstanding(batch, -1)
```
Comment (cjx0709) [inline python/sgl_jax/srt/disaggregation/pathways_scheduler.py]: Could we decrement `outstanding` only after the item is fully handled? If the capacity check below defers it, the item remains in `_pd_defer` but is already counted as drained. An abort in this window can free the P-side request, and the deferred item may later retry with stale slots.

File: python/sgl_jax/srt/disaggregation/pathways_scheduler.py [hunk 2]
```diff
@@ -955,6 +1075,15 @@ def _pd_maybe_push_prefill(self) -> None:
...
         for _off in range(n_p):
             i = (self._pd_next_p + _off) % n_p
+            # Aborted chunked owner with no batch in flight: finalize here so
+            # its P KV frees and no chunk N+1 is ever built (#1486). Owners
+            # with a batch still in flight finalize on that batch's drain.
+            for dp, owner in enumerate(self.p_chunked_reqs[i]):
+                if owner is None:
+                    continue
+                entry = self._pd_inflight.get(owner.rid)
+                if entry is not None and entry.pending_abort and entry.outstanding == 0:
```
Comment (cjx0709) [inline python/sgl_jax/srt/disaggregation/pathways_scheduler.py]: Could we skip building a continuation when `entry.pending_abort` is set, even if `outstanding > 0`? The P thread clears `_pd_chunk_pending` before the result is drained, so an abort in this window can still enqueue chunk N+1 and do extra work after cancellation.

File: python/sgl_jax/srt/disaggregation/pathways_scheduler.py [hunk 3]
```diff
@@ -869,18 +912,97 @@ def _pd_req_kv_tokens(self, r) -> int:
...
+                any(not q.empty() for q in self._pd_prefill_qs)
+                or any(self._pd_chunk_pending)
+                or not self._pd_ready_q.empty()
+                or bool(self._pd_defer)
+                or any(e.outstanding > 0 for e in self._pd_inflight.values())
+            )
+            if not busy:
+                return
+            time.sleep(0.01)
+        logger.warning(
```
Comment (cjx0709) [inline python/sgl_jax/srt/disaggregation/pathways_scheduler.py]: Could we fail closed when quiesce times out instead of only logging a warning? A late P result may still arrive after retract requeues the request, recreating the race this drain is intended to prevent.

File: python/sgl_jax/srt/managers/scheduler.py [hunk 4]
```diff
@@ -2794,6 +2801,13 @@ def pause_generation(self, recv_req: PauseGenerationReqInput):
         consumed = self._process_pending_chunked_aborts()
         self._retire_chunked_req_batch_owners(consumed)
 
+        # Pathways single-process PD: drain the async P pipeline as well so a
+        # late prefill result cannot merge into running_batch alongside a
+        # requeued copy of the same request after retract (#1486). No-op
+        # unless pathways PD is on.
+        if getattr(self, "_pd_inflight", None) is not None:
```
Comment (cjx0709) [inline python/sgl_jax/srt/managers/scheduler.py:2808]: Could retract also clear or requeue idle owners in `p_chunked_reqs`? With Pathways this helper is a no-op, so a request between chunks can remain only in the P pool and resume from its old prefix instead of being recomputed.

File: python/sgl_jax/srt/disaggregation/pathways_scheduler.py (hunk 1 above)
Comment (zorrofox) [reply python/sgl_jax/srt/disaggregation/pathways_scheduler.py]: Good catch — this was real: the decrement ran on every pop, so a deferred item's second pass would floor the count at zero while the item was still parked, and an abort in that window could finalize the req and free the P slots under it. Fixed in 6a71dbfc by moving the decrement past the admission gate (and into the failure branch), so a batch is only counted as drained once the drain actually commits to consuming it.

File: python/sgl_jax/srt/disaggregation/pathways_scheduler.py (hunk 2 above)
Comment (zorrofox) [reply python/sgl_jax/srt/disaggregation/pathways_scheduler.py]: Agreed — the P thread clears `_pd_chunk_pending` before the result is drained, so that window existed. Fixed in 6a71dbfc: the build loop now skips a slice entirely while its chunked owner has `pending_abort` with a batch still in flight; the owner is finalized on that batch's drain and no continuation is ever built.

File: python/sgl_jax/srt/disaggregation/pathways_scheduler.py (hunk 3 above)
Comment (zorrofox) [reply python/sgl_jax/srt/disaggregation/pathways_scheduler.py]: Agreed, and fail-closed is the only safe direction here: the stranded P pages cannot be reclaimed while the prefill thread may still be executing on them, so neither "free and requeue" nor "proceed and hope" is sound. 6a71dbfc makes `_pd_quiesce` block until the pipeline actually drains, logging a periodic warning (cold-shape compiles on the P thread can legitimately take >30s, which is exactly when the old timeout would have fired). If the P thread dies without reporting, the failure path still enqueues a `result=None` item, so the drain completes through the failure branch.

File: python/sgl_jax/srt/managers/scheduler.py (hunk 4 above)
Comment (zorrofox) [reply python/sgl_jax/srt/managers/scheduler.py:2808]: Right — `_retract_parked_chunked_reqs` early-returns under pathways PD and the parked owner would have resumed on pre-retract KV (stale if the retract preceded a weight update). 6a71dbfc adds `_pd_retract_chunked_owners`, called from the retract branch after the quiesce: frees the P-side KV and req slot, drops the registry/admission accounting, resets the req and requeues it so it recomputes from scratch. Unit test covers the requeue-not-abort semantics.

**Review by cjx0709 — APPROVED** [review]
LGTM
